python/llm/src/bigdl/llm/transformers/models/chatglm2.py

```diff
@@ -119,19 +119,56 @@ def chatglm2_attention_forward_8eb45c(
+
+    if kv_cache is not None:  # rest tokens
+        position_ids = torch.tensor([[kv_cache[0].shape[0]]], dtype=torch.long).to(query_layer.device)
+    else:
+        position_ids = torch.tensor([range(0, cur_length)], dtype=torch.long).to(query_layer.device)
+    # print("-----attention position ids:", position_ids)
+
     # apply relative positional encoding (rotary embedding)
+    use_fuse_rope = query_layer.device.type == "xpu"
+    use_fuse_rope = use_fuse_rope and not (self.training and query_layer.requires_grad)
+    # use_fuse_rope = False
     if rotary_pos_emb is not None:
-        query_layer = apply_rotary_pos_emb(query_layer, rotary_pos_emb)
-        key_layer = apply_rotary_pos_emb(key_layer, rotary_pos_emb)
-
-    cur_length, batch_size = query_layer.shape[0], query_layer.shape[1]
+        if use_fuse_rope:
+            rot_dim = rotary_pos_emb.shape[-2] * 2
+            query_layer = query_layer.permute(1, 2, 0, 3)
+            query_layer, query_layer_pass = query_layer[..., :rot_dim], query_layer[..., rot_dim:]
+            key_layer = key_layer.permute(1, 2, 0, 3)
+            key_layer, key_layer_pass = key_layer[..., :rot_dim], key_layer[..., rot_dim:]
+            dummy_q = torch.empty(query_layer.shape, dtype=query_layer.dtype, device=query_layer.device)
+            dummy_k = torch.empty(key_layer.shape, dtype=key_layer.dtype, device=key_layer.device)
+            _, query_layer = apply_rotary_pos_emb_no_cache_xpu(dummy_q,
+                                                               query_layer,
+                                                               position_ids,
+                                                               "llama")
+            _, key_layer = apply_rotary_pos_emb_no_cache_xpu(dummy_k,
+                                                             key_layer,
+                                                             position_ids,
+                                                             "llama")
+            if query_layer_pass.shape[-1] > 0:
+                query_layer = torch.cat((query_layer, query_layer_pass), dim=-1)
+            query_layer = query_layer.permute(2, 0, 1, 3)
+            if key_layer_pass.shape[-1] > 0:
+                key_layer = torch.cat((key_layer, key_layer_pass), dim=-1)
+            key_layer = key_layer.permute(2, 0, 1, 3)
+        else:
+            query_layer = apply_rotary_pos_emb_chatglm(query_layer, rotary_pos_emb)
+            key_layer = apply_rotary_pos_emb_chatglm(key_layer, rotary_pos_emb)
 
     if self.multi_query_attention:
         key_length = key_layer.size(0)
         query_group_size = self.num_attention_heads_per_partition // \
             self.num_multi_query_groups_per_partition
+        # if rotary_pos_emb is not None and use_fuse_rope:
+        #     key_layer = key_layer.unsqueeze(-3)
+        # else:
+        #     key_layer = key_layer.permute(1, 2, 0, 3).unsqueeze(-3)  # [bs, nh/k, sl, hn]
```

Comment:
In theory this permute can be merged with xpu kernel's output shape, but don't know why it gets the error:
```
  File "/home/arda/anaconda3/envs/kai-llm-pip/lib/python3.9/site-packages/bigdl/llm/transformers/models/chatglm2.py", line 193, in chatglm2_attention_forward_8eb45c
    key_layer = key_layer.contiguous().view((batch_size,
RuntimeError: shape '[1, 32, 1, 128]' is invalid for input of size 53248
```